### DISCUSSION TITLE
Replace all string on a CST node and use a dictionary

### DISCUSSION THREAD

**Discussion by KFlash** [kataw/kataw#79] [Ideas]
It could be an idea to replace all string on a CST node and use a dictionary  lookup if we want an even faster parser.

For example 

```ts
 {
            "kind": 120,
            "expression": {
                "kind": 134299649,
                "text": "a",
                "rawText": "a",
                "flags": 96,
                "start": 0,
                "end": 1
            },
            "flags": 16,
            "start": 0,
            "end": 1
        }
```

would be

```ts
 {
           1: 120,
            2: {
                3: 134299649,
                4: "a",
                5: "a",
                6: 96,
                7: 0,
                8: 1
            },
            6: 16,
            7: 0,
            8: 1
        }
```

I'm not sure what's best.

@aladdin-add What do you prefer? The latter?

**Comment by aladdin-add** [comment]
I prefer the former for better readability, if there is not a very big perf diff.

**Reply by KFlash** [reply]
I think we will only gain reduced CST size, not much on performance. We could instead try to "compress" existing CST nodes into a subset of existing ECMA syntax.  For example `{x: x}` could be `{x}` instead, and then return an identifier instead of a CST node as we do now.   `false` could be  `0` etc.

Do you want to look into that?

**Comment by KFlash** [comment]
if we try that with for example  `false` vs  `0` we could skip the boolean literal CST node and use numeric literal instead and add a new  `NodeFlag` telling this is an  `boolean`

**Comment by KFlash** [comment]
@aladdin-add As of now I found some other ways to reduce the CST size. 
E.g. ArrayBindingPattern is a `SingleNameBinding` by default or if it has an initializer or something like this case

`let [x?: string = y]: string[] = z`

it will turn into `BindingElement`.  It conforms to the ECMA specs
